File: front/web/lib/policy.ts
/**
 * The Policy DSL — the Workbench /policy projection source (AIDOS step S09).
 *
 * AUTHORIZATION-AS-ARTIFACT (KRD §24.4, §93): a policy is a TYPED rule tree —
 * combinators all/any/not over comparison leaves (eq/gt/lt) and predicates
 * exists/matches on a $-rooted selector — with a scope (RESOURCE/OPERATION/
 * ENTITY/FIELD) and an effect (ALLOW/DENY), stored content-addressed in
 * kernel.policy, never free code. This module holds the DECLARED description of
 * that shape (the rule kinds, the scopes — field-for-field with
 * back/kernel/policy), the §93 canPlaceOrder anchor, and a tiny deterministic
 * evaluator that mirrors Go's policy.Eval for sample contexts — so the /policy
 * panel renders the rule tree and its ALLOW/DENY verdict before the live
 * kernel.policy SELECT wiring lands.
 *
 * DETERMINISM-FIRST (CLAUDE.md §6/§8): a static registry + a pure evaluator (no
 * clock, no rng, no I/O). The reproducibility mirror lib/policy.test.ts pins the
 * rule-kind set, the scope set, and the sample verdicts against the Go invariants
 * (canPlaceOrder ALLOW iff the three §93 conditions, else DENY).
 *
 * READ-ONLY (CLAUDE.md §7 ui-completeness): /policy is a visualization; it
 * projects kernel.policy (SELECT-only role), it does not write truth and exposes
 * no capability — a Policy AST is written only by the aidos CLI through an
 * approved ChangeSet. There is no headless capability hidden here; read-only is
 * correct.
 */

/** The eight rule-node kinds (matches policy.go ruleKinds, canonical order). */
export const RULE_KINDS = [
	"all",
	"any",
	"not",
	"eq",
	"gt",
	"lt",
	"exists",
	"matches",
] as const;
export type RuleKind = (typeof RULE_KINDS)[number];

/** The four scopes (matches policy.go scopes, canonical order). */
export const SCOPES = ["RESOURCE", "OPERATION", "ENTITY", "FIELD"] as const;
export type Scope = (typeof SCOPES)[number];

/** The effect of a policy: ALLOW gates, DENY blocks (KRD §24.4). */
export type Effect = "ALLOW" | "DENY";
/** A total decision (KRD §93: every ALLOW passes; any DENY blocks). */
export type Decision = "ALLOW" | "DENY";

/** An operand: a $-rooted selector or a literal value. */
export type Operand =
	| { kind: "sel"; path: string }
	| { kind: "lit"; value: string | number | boolean | null };

/** A rule-tree node — a discriminated union, the closed grammar. */
export type Rule =
	| { kind: "all"; children: Rule[] }
	| { kind: "any"; children: Rule[] }
	| { kind: "not"; child: Rule }
	| { kind: "eq"; left: Operand; right: Operand }
	| { kind: "gt"; left: Operand; right: Operand }
	| { kind: "lt"; left: Operand; right: Operand }
	| { kind: "exists"; sel: string }
	| { kind: "matches"; sel: string; pattern: string };

/** A typed Policy (matches the Go Policy struct). */
export interface Policy {
	name: string;
	scope: Scope;
	target: string;
	rule: Rule;
	effect: Effect;
}

/** Is k one of the eight rule kinds? (mirrors policy.IsRuleKind). */
export function isRuleKind(k: string): k is RuleKind {
	return (RULE_KINDS as readonly string[]).includes(k);
}

/** Is s one of the four scopes? (mirrors policy.IsScope). */
export function isScope(s: string): s is Scope {
	return (SCOPES as readonly string[]).includes(s);
}

/** A $-rooted ctx value tree (mirrors the Go Ctx data under "$"). */
export type CtxData = Record<string, unknown>;

/**
 * The KRD §93 anchor policy canPlaceOrder, VERBATIM (no invented rule):
 *   scope OPERATION "createOrder", effect ALLOW,
 *   rule all([ exists($.auth.user),
 *              eq($.cart.userId, $.auth.user.id),
 *              gt($.cart.items.length, 0) ]).
 */
export const CAN_PLACE_ORDER: Policy = {
	name: "canPlaceOrder",
	scope: "OPERATION",
	target: "createOrder",
	effect: "ALLOW",
	rule: {
		kind: "all",
		children: [
			{ kind: "exists", sel: "$.auth.user" },
			{
				kind: "eq",
				left: { kind: "sel", path: "$.cart.userId" },
				right: { kind: "sel", path: "$.auth.user.id" },
			},
			{
				kind: "gt",
				left: { kind: "sel", path: "$.cart.items.length" },
				right: { kind: "lit", value: 0 },
			},
		],
	},
};

/** Resolve a $-rooted dotted path; ".length" yields an array/string length. */
function resolveSelector(path: string, root: unknown): unknown {
	if (!path.startsWith("$")) return undefined;
	const segs = path.split(".").slice(1);
	let cur: unknown = root;
	for (const s of segs) {
		if (s === "length") {
			if (Array.isArray(cur) || typeof cur === "string") return cur.length;
			return undefined;
		}
		if (typeof cur !== "object" || cur === null) return undefined;
		cur = (cur as Record<string, unknown>)[s];
	}
	return cur;
}

function resolveOperand(o: Operand, root: unknown): unknown {
	return o.kind === "sel" ? resolveSelector(o.path, root) : o.value;
}

function asNumber(v: unknown): number | undefined {
	return typeof v === "number" ? v : undefined;
}

/**
 * holds reports whether a rule tree holds under the ctx root. Mirrors Go's
 * policy.Holds: a failed selector / type mismatch makes a leaf NOT hold (totality),
 * never throws.
 */
export function holds(r: Rule, root: unknown): boolean {
	switch (r.kind) {
		case "all":
			return r.children.every((c) => holds(c, root));
		case "any":
			return r.children.some((c) => holds(c, root));
		case "not":
			return !holds(r.child, root);
		case "eq": {
			const l = resolveOperand(r.left, root);
			const rt = resolveOperand(r.right, root);
			const ln = asNumber(l);
			const rn = asNumber(rt);
			if (ln !== undefined && rn !== undefined) return ln === rn;
			return l === rt;
		}
		case "gt":
		case "lt": {
			const ln = asNumber(resolveOperand(r.left, root));
			const rn = asNumber(resolveOperand(r.right, root));
			if (ln === undefined || rn === undefined) return false;
			return r.kind === "gt" ? ln > rn : ln < rn;
		}
		case "exists": {
			const v = resolveSelector(r.sel, root);
			return v !== undefined && v !== null;
		}
		case "matches": {
			const v = resolveSelector(r.sel, root);
			if (typeof v !== "string") return false;
			try {
				return new RegExp(r.pattern).test(v);
			} catch {
				return false;
			}
		}
	}
}

/**
 * evaluate applies the §93 combination law via the policy's effect gate, returning
 * a total Decision. Mirrors Go's policy.Eval: ALLOW iff the rule holds (for an
 * ALLOW gate), else DENY; a DENY gate blocks iff the rule holds, else ALLOW.
 * The ctx is the value UNDER "$" (callers pass the root tree directly).
 */
export function evaluate(p: Policy, root: CtxData): Decision {
	const held = holds(p.rule, root);
	if (p.effect === "ALLOW") return held ? "ALLOW" : "DENY";
	return held ? "DENY" : "ALLOW";
}

/**
 * canonicalize serializes a Policy to the SAME canonical JSON bytes as Go's
 * policy.Canonicalize (keys statically sorted at every level, arrays in order, no
 * insignificant whitespace). It is the byte source of the content-address, so the
 * id the panel shows matches the kernel.policy id Go computes. Determinism-first:
 * a pure string transform, no I/O.
 */
export function canonicalize(p: Policy): string {
	const operand = (o: Operand): string =>
		o.kind === "sel"
			? `{"kind":"sel","path":${JSON.stringify(o.path)}}`
			: `{"kind":"lit","value":${JSON.stringify(o.value)}}`;
	const rule = (r: Rule): string => {
		switch (r.kind) {
			case "all":
			case "any":
				return `{"children":[${r.children.map(rule).join(",")}],"kind":${JSON.stringify(r.kind)}}`;
			case "not":
				return `{"child":${rule(r.child)},"kind":"not"}`;
			case "eq":
			case "gt":
			case "lt":
				return `{"kind":"compare","left":${operand(r.left)},"op":${JSON.stringify(r.kind)},"right":${operand(r.right)}}`;
			case "exists":
				return `{"kind":"exists","sel":${JSON.stringify(r.sel)}}`;
			case "matches":
				return `{"kind":"matches","pattern":${JSON.stringify(r.pattern)},"sel":${JSON.stringify(r.sel)}}`;
		}
	};
	return `{"effect":${JSON.stringify(p.effect)},"kind":"policy","name":${JSON.stringify(p.name)},"rule":${rule(p.rule)},"scope":${JSON.stringify(p.scope)},"target":${JSON.stringify(p.target)}}`;
}

/** One seeded authorization sample: a ctx and the verdict it should produce. */
export interface PolicySample {
	id: string;
	/** FR-first label (ADR 0011) — what this context represents. */
	role: string;
	ctx: CtxData;
}

/**
 * The seeded sample contexts the /policy panel evaluates against canPlaceOrder.
 * They are the fixture-mirror rows, so the panel shows exactly what the Go fixture
 * proves: an authed user with a matching non-empty cart → ALLOW; a broken
 * condition → DENY.
 */
export const SAMPLES: readonly PolicySample[] = [
	{
		id: "allow-authed-matching-cart",
		role: "Un utilisateur connecté, propriétaire d'un panier non vide → ALLOW.",
		ctx: {
			auth: { user: { id: "u1" } },
			cart: { userId: "u1", items: [{ id: 1 }] },
		},
	},
	{
		id: "deny-no-auth",
		role: "Aucun utilisateur connecté → DENY (la première règle, exists, échoue).",
		ctx: { auth: {}, cart: { userId: "u1", items: [{ id: 1 }] } },
	},
	{
		id: "deny-empty-cart",
		role: "Panier vide → DENY (la règle gt($.cart.items.length, 0) échoue).",
		ctx: { auth: { user: { id: "u1" } }, cart: { userId: "u1", items: [] } },
	},
] as const;

/** Look up a seeded sample by id (throws if absent — no non-null assertion). */
export function sampleById(id: string): PolicySample {
	const s = SAMPLES.find((x) => x.id === id);
	if (!s) throw new Error(`no policy sample with id "${id}"`);
	return s;
}

/** The decision of a sample against canPlaceOrder, as a display string. */
export function sampleDecision(s: PolicySample): Decision {
	return evaluate(CAN_PLACE_ORDER, s.ctx);
}

/** A flat node list of a rule tree (depth-first), for the tree rendering. */
export interface FlatRule {
	/** Stable depth-first position, unique within one flattened tree (React key). */
	id: number;
	depth: number;
	kind: RuleKind;
	/** A one-line, human label of the node. */
	label: string;
}

function operandLabel(o: Operand): string {
	return o.kind === "sel" ? o.path : JSON.stringify(o.value);
}

export function flatten(rule: Rule): FlatRule[] {
	const out: FlatRule[] = [];
	let next = 0;
	const walk = (n: Rule, depth: number): void => {
		const id = next++;
		switch (n.kind) {
			case "all":
			case "any":
				out.push({ id, depth, kind: n.kind, label: `${n.kind}([…])` });
				for (const c of n.children) walk(c, depth + 1);
				return;
			case "not":
				out.push({ id, depth, kind: "not", label: "not(…)" });
				walk(n.child, depth + 1);
				return;
			case "eq":
			case "gt":
			case "lt":
				out.push({
					id,
					depth,
					kind: n.kind,
					label: `${operandLabel(n.left)} ${n.kind} ${operandLabel(n.right)}`,
				});
				return;
			case "exists":
				out.push({ id, depth, kind: "exists", label: `exists(${n.sel})` });
				return;
			case "matches":
				out.push({
					id,
					depth,
					kind: "matches",
					label: `matches(${n.sel}, ${JSON.stringify(n.pattern)})`,
				});
				return;
		}
	};
	walk(rule, 0);
	return out;
}

File: back/kernel/policy/policy.go
// Package policy is the AIDOS Policy DSL (KRD §24.4, §93): a content-addressed
// AST for a recursive ALLOW/DENY rule tree, plus a PURE evaluator. A policy is a
// SOURCE layer above the waterline (KRD §21, back/kernel/CONTEXT.md) — human
// truth, out of AI reach — that answers one question: "is this authorized?".
//
// THE SHAPE (KRD §24.4). A Policy is {name, scope, rule, effect}:
//   - scope ∈ {RESOURCE, OPERATION, ENTITY, FIELD} with a target name — WHERE the
//     policy applies (the §93 anchor scopes OPERATION "createOrder");
//   - rule  — a recursive tree: combinators all / any / not over comparison leaves
//     (eq / gt / lt / …) and predicates exists / matches on a $-rooted selector;
//   - effect ∈ {ALLOW, DENY} — how the rule's truth maps to a Decision.
//
// THE COMBINATION LAW (KRD §93: "tout ALLOW passe ; n'importe quel DENY bloque").
// For a SINGLE policy, the effect is the gate: an ALLOW policy authorizes iff its
// rule HOLDS (else DENY); a DENY policy blocks iff its rule HOLDS (else ALLOW —
// it abstains). Across a POLICY SET (a later tooth, not wired here), every ALLOW
// must hold and any DENY that holds blocks — the §93 law at the set level.
//
// THE WALL (CLAUDE.md §2). This package is PURE — no DB calls, no I/O, no clock,
// no RNG. The kernel.policy truth table is SELECT-only to the agent role; a
// Policy AST is written only by the aidos CLI role through an approved ChangeSet.
// Eval is a deterministic function of (policy, ctx): same inputs ⇒ same Decision
// (CLAUDE.md §6 determinism-first).
//
// REUSE, DON'T REINVENT (ADR 0007). The selector (exists / matches / a leaf's
// operands) reuses the SAME $-rooted dotted-path semantics proved in
// back/kernel/expr (resolveRef: ".length" yields a collection's length); we do
// NOT embed a general JSON-path engine — that would be a free-code escape, the
// very thing §24 forbids. Hashing reuses records.Hash/Canonicalize (one address
// space across stores). No new ADR: this freezes a new artifact (the Policy AST
// + the ALLOW/DENY law); it shifts no prior contract.
package policy

// Scope places a policy: WHERE it applies (KRD §24.4). The closed set is
// {RESOURCE, OPERATION, ENTITY, FIELD}; New rejects any other value at
// construction so a scope is never invented.
type Scope string

const (
	ScopeResource  Scope = "RESOURCE"
	ScopeOperation Scope = "OPERATION"
	ScopeEntity    Scope = "ENTITY"
	ScopeField     Scope = "FIELD"
)

// scopes is the closed set of scope values, in canonical order.
var scopes = []Scope{ScopeResource, ScopeOperation, ScopeEntity, ScopeField}

// IsScope reports whether s is one of the four Policy scopes. Exposed so the
// Workbench panel and the property mirror never invent a scope.
func IsScope(s string) bool {
	for _, sc := range scopes {
		if string(sc) == s {
			return true
		}
	}
	return false
}

// Scopes returns the four scope values in canonical order.
func Scopes() []Scope { return append([]Scope(nil), scopes...) }

// Effect maps a rule's truth to a Decision (KRD §24.4): ALLOW gates, DENY blocks.
type Effect string

const (
	EffectAllow Effect = "ALLOW"
	EffectDeny  Effect = "DENY"
)

// Decision is the total result of evaluating a policy: ALLOW or DENY. There is no
// third value — Eval is total (the §93 "every ALLOW passes; any DENY blocks" law
// always yields a verdict).
type Decision string

const (
	ALLOW Decision = "ALLOW"
	DENY  Decision = "DENY"
)

// Policy is one content-addressed authorization rule tree (KRD §24.4, §93).
// {Name, Scope, Target, Rule, Effect}. Target is the scoped name (the §93 anchor:
// scope OPERATION, target "createOrder").
type Policy struct {
	Name   string
	Scope  Scope
	Target string
	Rule   Rule
	Effect Effect
}

// New constructs a Policy. It is a plain constructor (no validation panic): an
// out-of-set scope or effect is simply held as-is and surfaced by Canonicalize /
// the validator later. Callers building the §93 anchor use CanPlaceOrder().
func New(name string, scope Scope, target string, rule Rule, effect Effect) Policy {
	return Policy{Name: name, Scope: scope, Target: target, Rule: rule, Effect: effect}
}

// ── The rule tree (a sealed sum type) ────────────────────────────────────────

// RuleKind discriminates a Rule node. The closed set is the combinators (all /
// any / not), the comparison leaves (eq / gt / lt), and the selector predicates
// (exists / matches). Nothing is added beyond what KRD §24.4 / §93 names.
type RuleKind string

const (
	KindAll     RuleKind = "all"     // combinator: every child holds
	KindAny     RuleKind = "any"     // combinator: some child holds
	KindNot     RuleKind = "not"     // combinator: negation
	KindEq      RuleKind = "eq"      // leaf: left == right
	KindGt      RuleKind = "gt"      // leaf: left > right (numbers)
	KindLt      RuleKind = "lt"      // leaf: left < right (numbers)
	KindExists  RuleKind = "exists"  // predicate: a selector resolves to a value
	KindMatches RuleKind = "matches" // predicate: a selector's string matches a pattern
)

var ruleKinds = []RuleKind{KindAll, KindAny, KindNot, KindEq, KindGt, KindLt, KindExists, KindMatches}

// IsRuleKind reports whether k is one of the eight rule-node kinds.
func IsRuleKind(k string) bool {
	for _, rk := range ruleKinds {
		if string(rk) == k {
			return true
		}
	}
	return false
}

// RuleKinds returns the eight rule kinds in canonical order.
func RuleKinds() []RuleKind { return append([]RuleKind(nil), ruleKinds...) }

// Rule is one node of the rule tree. Sealed interface: only the concrete node
// types in this package implement it (the closed grammar — no free-code escape).
type Rule interface {
	Kind() RuleKind
	isRule()
}

// AllNode holds iff every child holds (the ∧ combinator). Empty all([]) holds
// (vacuous truth), matching boolean conjunction over the empty set.
type AllNode struct{ Children []Rule }

func (AllNode) Kind() RuleKind { return KindAll }
func (AllNode) isRule()        {}

// AnyNode holds iff some child holds (the ∨ combinator). Empty any([]) fails.
type AnyNode struct{ Children []Rule }

func (AnyNode) Kind() RuleKind { return KindAny }
func (AnyNode) isRule()        {}

// NotNode holds iff its child does not (negation; an involution).
type NotNode struct{ Child Rule }

func (NotNode) Kind() RuleKind { return KindNot }
func (NotNode) isRule()        {}

// CompareNode is a comparison leaf: Left <op> Right, where op ∈ {eq, gt, lt}.
// Operands are Operands (a selector or a literal).
type CompareNode struct {
	Op    RuleKind // KindEq / KindGt / KindLt
	Left  Operand
	Right Operand
}

func (n CompareNode) Kind() RuleKind { return n.Op }
func (CompareNode) isRule()          {}

// ExistsNode holds iff its selector resolves to a non-null value in the ctx.
type ExistsNode struct{ Sel string }

func (ExistsNode) Kind() RuleKind { return KindExists }
func (ExistsNode) isRule()        {}

// MatchesNode holds iff the selector resolves to a string matching Pattern (a
// Go regexp). Compilation failure ⇒ the predicate fails (never a panic).
type MatchesNode struct {
	Sel     string
	Pattern string
}

func (MatchesNode) Kind() RuleKind { return KindMatches }
func (MatchesNode) isRule()        {}

// ── Operands (a comparison leaf's left/right) ────────────────────────────────

// OperandKind discriminates an operand: a $-rooted selector or a literal value.
type OperandKind string

const (
	OperandSelector OperandKind = "sel"
	OperandLiteral  OperandKind = "lit"
)

// Operand is one side of a comparison leaf. Sealed to this package.
type Operand interface {
	operandKind() OperandKind
	isOperand()
}

// SelectorOperand is a $-rooted dotted path resolved against the ctx (the same
// semantics as expr: ".length" yields a collection's length).
type SelectorOperand struct{ Path string }

func (SelectorOperand) operandKind() OperandKind { return OperandSelector }
func (SelectorOperand) isOperand()               {}

// LiteralOperand is a constant value: string, float64 (a JSON number), bool, nil.
type LiteralOperand struct{ Value any }

func (LiteralOperand) operandKind() OperandKind { return OperandLiteral }
func (LiteralOperand) isOperand()               {}

// ── Constructors (used by callers / mirrors building a rule tree in Go) ───────

// All builds an all-combinator (every child holds).
func All(children ...Rule) Rule { return AllNode{Children: children} }

// Any builds an any-combinator (some child holds).
func Any(children ...Rule) Rule { return AnyNode{Children: children} }

// Not builds a negation.
func Not(child Rule) Rule { return NotNode{Child: child} }

// Eq builds an equality leaf (left == right).
func Eq(left, right Operand) Rule { return CompareNode{Op: KindEq, Left: left, Right: right} }

// Gt builds a greater-than leaf (left > right, numbers).
func Gt(left, right Operand) Rule { return CompareNode{Op: KindGt, Left: left, Right: right} }

// Lt builds a less-than leaf (left < right, numbers).
func Lt(left, right Operand) Rule { return CompareNode{Op: KindLt, Left: left, Right: right} }

// Exists builds an exists predicate over a selector.
func Exists(sel string) Rule { return ExistsNode{Sel: sel} }

// Matches builds a matches predicate (selector string ~ pattern).
func Matches(sel, pattern string) Rule { return MatchesNode{Sel: sel, Pattern: pattern} }

// Selector builds a selector operand ($-rooted path).
func Selector(path string) Operand { return SelectorOperand{Path: path} }

// LitV builds a literal operand, normalizing Go int → float64 so the canonical
// encoding matches a decoded JSON number (one number type, like expr.Lit).
func LitV(v any) Operand {
	switch n := v.(type) {
	case int:
		return LiteralOperand{Value: float64(n)}
	case int64:
		return LiteralOperand{Value: float64(n)}
	default:
		return LiteralOperand{Value: v}
	}
}
